src/profile-logic/transforms.js

```diff
@@ -21,35 +22,69 @@ import type {
 } from '../types/profile';
 import type { CallNodePath } from '../types/profile-derived';
 import type { ImplementationFilter } from '../types/actions';
-import type { Transform, TransformStack } from '../types/transforms';
+import type {
+  Transform,
+  TransformType,
+  TransformStack,
+} from '../types/transforms';
 
 /**
  * This file contains the functions and logic for working with and applying transforms
  * to profile data.
  */
 
+// Create mappings from a transform name, to a url-friendly short name.
+export const TRANSFORM_TO_SHORT_KEY: { [TransformType]: string } = {};
+export const SHORT_KEY_TO_TRANSFORM: { [string]: TransformType } = {};
+[
+  'focus-subtree',
+  'focus-function',
+  'merge-call-node',
+  'merge-function',
+  'drop-function',
+  'collapse-resource',
+  'collapse-direct-recursion',
+  'collapse-function-subtree',
+].forEach((transform: TransformType) => {
+  // This is kind of an awkward switch, but it ensures we've exhaustively checked that
+  // we have a mapping for every transform.
```

Comment:
Are you sure this works as you think? I believe you can add or remove something from the array you're iterating with without flow warning you...

[This seems to work better](https://flow.org/try/#0C4TwDgpgBAKgTgQwHYGcBmB7OBbG4IpQC8AUFFAD5QDkmAxgK4oC0KDARsHBBNWZTXpNmaBkjrAAlhiR9yVatghwA5hGZ0EAGy3MkGACa9+CpavWjxUmXIHUDcDGBFiJ02SZp0MOhGBTq3CgYDHB0xvJePlp+AcwGktwSzEmhKO7UANwkJN6owLAASgCCAHIAygBiAPKFALIA+jDVDeUAErUwDQDSAKIAmgBcUADeANoA1hAgw-DI6Fi4+CgAusMAJN3TKAA8oJAYaFDtnT0DTS0wJRU19QB8AL7Eo-y0GIwsbJzcvMO01AAaV5CFiWNw2P5oNCA15mNQabS6fRGah-bB0WRA8iKZTwsHWWRo6FYmgOJwuKwZP4GYmvby+fyBAghMK-LxwGHY+kxRnxRIQZKpODpCE0JIwh7ZXIyFAFE6FLp9foXJrXKq1OrDcZTGZQdYANW0DAIe3whyKZXV9RV8sVAzua1giFQmBweEgKCeRBe5DQkPewi+XB4nKgUP9Hwp4Mx-HRSDRuPUmh0ekMvBJ2D9NDhFlcBNDNOpjmc+IyJLocD+3NiTOCoXCoaSVeiNb5SWAKQFaTLJElQA)

```javascript
type TransformTypes = // comes from types/transforms.js
  | 'focus-subtree'
  | 'focus-function'
  | 'merge-call-node'
  | 'merge-function'
  | 'drop-function'
  | 'collapse-resource'
  | 'collapse-direct-recursion';

const TRANSFORM_TO_SHORT_KEY: {[key: TransformTypes]: $Keys<typeof SHORT_KEY_TO_TRANSFORM>} = {
  'focus-subtree': 'f',
  'focus-function': 'ff',
  'merge-call-node': 'mcn',
  'merge-function': 'mf',
  'drop-function': 'df',
  'collapse-resource': 'cr',
  'collapse-direct-recursion': 'rec',
};

const SHORT_KEY_TO_TRANSFORM: {[key: $Values<typeof TRANSFORM_TO_SHORT_KEY>]: TransformTypes} = {
  f: 'focus-subtree',
  ff: 'focus-function',
  mcn: 'merge-call-node',
  mf: 'merge-function',
  df: 'drop-function',
  cr: 'collapse-resource',
  rec: 'collapse-direct-recursion',
};
```

Adding a new transform type wouldn't force you add it to these structures (I don't know how to do it, I think we'd need to check that `$Keys<typeof TRANSFORM_TO_SHORT_KEY> === TransformTypes` but I don't know if that's possible), but your current code doesn't either. But this keeps the 2 structures synced in a type-safe way, with a much simpler code.

Reply:
Mainly I wanted it to throw an error when I add a new type, so that I know to go and hit that code and update it. It's not 100% foolproof. I tried yours and it's not giving me errors when I add a new transform type.

With the switch it handles the unexpected case:
<img width="890" alt="untitled" src="https://user-images.githubusercontent.com/1588648/35241247-7e5e8dae-ff7b-11e7-83cf-c709e2eb6443.png">

Reply:
I see, this is because you added the `TransformType` type to the argument (it wasn't there in the first commit), and Flow doesn't understand that not all types will be passed.
OK then.